### DISCUSSION TITLE
tiffsave ndpi color differences

### DISCUSSION THREAD

**Discussion by alexlsa** [libvips/libvips#3780] [General]
Hello,

When working with `.ndpi` WSI, there is some color differences, it is hard to see differencess but it is there

This is my command:

`vips tiffsave x.ndpi x.tif --pyramid --tile --compression jpeg `

I think there is alpha channel differencess but I am not sure

How can I correct this ?

.ndpi image
![orijinal_ndpi](https://github.com/libvips/libvips/assets/148692535/39a60f00-806f-4eaa-9fd4-86838039e929)

.tif image
![vips_tiffsave](https://github.com/libvips/libvips/assets/148692535/55d2a27e-ce3b-4453-9a51-c68b821d040a)

**Comment by jcupitt** [comment]
Hi @alexlsa,

It depends on many factors: the libvips version, the openslide version, the image viewer you are using for the ndpi images, the scanner model, etc., so you'll need to post some more information. 

The big recent change was colour profile support in openslide4 and libvips 8.15. I would make sure you are using both those versions.

The vipsdisp binary on flathub can be useful for testing:

https://flathub.org/apps/org.libvips.vipsdisp

That's openslide4 + libvips 8.15. Use the save-as menu to make a TIFF.

**Comment by alexlsa** [comment]
Hi @jcupitt,

First of all thank you for very fast response

```
vips --version
vips-8.14.5
```

openslide is also inside vips. I am using openslide-python [viewer](https://github.com/openslide/openslide-python/tree/main/examples/deepzoom). I think, scanner is NanoZoomer S60.

openslide-python viewer has openslide4 version, and `.ndpi` image has no ICC profile

I also tried  with `vips-8.15` but it is same in openslide-python viewer

I installed [vipsdisp ](https://flathub.org/apps/org.libvips.vipsdisp )and I have looked both wsi, actually I could not realize there is or not difference but it say;

```
.ndpi image has "uchar, 4 bands, srgb"
.tif image has "10 pages, uchar, 3 bands, srgb"
```

I can also share original .ndpi file

**Reply by jcupitt** [reply]
Ooop, sorry, my mistake, openslide does not support icc profiles in NDPI files, so it won't make a difference.

openslide-python and libvips are using almost the same processing libraries, so it's hard to understand how they can look different. Maybe it's the JPEG encoding options? You could try:

```
vips tiffsave x.ndpi x.tif --pyramid --tile --compression jpeg --Q 90
```

**Reply by jcupitt** [reply]
> .ndpi image has "uchar, 4 bands, srgb"
> .tif image has "10 pages, uchar, 3 bands, srgb"

Yes, that's expected behaviour.

**Reply by alexlsa** [reply]
Also why is expected ? for example .ndpi 4 bands but .tif 3 bands

**Reply by jcupitt** [reply]
openslide always writes RGBA (ie. always sets an alpha), even when it's not needed. JPEG compression can only do RGB (no alpha), so the alpha is automatically flattened.

You can force openslide to only write RGB with the `--rgb` flag and save the flatten step, for example:

```
vips tiffsave x.ndpi[rgb] x.tif --pyramid --tile --compression jpeg --Q 90
```

You'll find it's noticeably quicker.

**Comment by alexlsa** [comment]
Yes .ndpi files has no icc profile (.ndpi viewer has ICC option) 

I tried jpeg --Q 90 but still same

Actually when looking in Qupath, I think there is no difference but I am not sure

This is drive [link ](https://drive.google.com/file/d/1aInyeJ6e6CX7IYmOVDZrUp0vwlinfZP0/view?usp=drive_link) is it possible to look at it ?

**Reply by jcupitt** [reply]
If they look the same in qupath I think everything is probably working OK.

**Reply by alexlsa** [reply]
Hi @jcupitt, I am sorry, I was wrong there is color differenceses in qupath also. 

I also converted via qupath export option but this time there are no differences with original .ndpi wsi

I also use `deflate` compression and I compared pixel values with PIL, they are equal, I do not know what is the problem 

please look at this images

original
![original](https://github.com/libvips/libvips/assets/148692535/ba0200c9-2552-42b6-9eb7-8c40159d182b)

deflate
![deflate_compress](https://github.com/libvips/libvips/assets/148692535/b620b8cf-0117-4d42-b764-33ace826eb8d)

original
![original_2](https://github.com/libvips/libvips/assets/148692535/367d9efd-b4bd-4ad6-8512-3b05a32a1d77)

deflate
![deflate_compress_2](https://github.com/libvips/libvips/assets/148692535/99af26f6-961d-4d6d-ab43-c7650df3eae6)

**Comment by jcupitt** [comment]
I had a quick look at your file, you have the extension `.tiff` on it, but it's really a `.ndpi`, is that right?

I loaded the file into qupath 0.4.4 and vipsdisp:

![image](https://github.com/libvips/libvips/assets/580843/4843c576-3451-4a52-ac6f-9a62320553d5)

And they look identical to me.

**Reply by jcupitt** [reply]
Ah, I found a difference:

http://www.rollthepotato.net/~john/qupath.png

http://www.rollthepotato.net/~john/vips.png

Load those two images into two browser tabs, then try flipping between them.

The vips one is from vipsdisp and is using openslide4. You can see it's positioning the camera tiles differently -- it's interpreting the extra metadata in the TIFF to align the tiles correctly.

QuPath 0.4.4 seems to be loading the file as a simple TIFF and not interpreting these extra fields, so the tiles are not in exactly the right place. QuPath 0.5 might do a better job. 

I don't see any colour differences though.

**Comment by alexlsa** [comment]
Thank you for your time @jcupitt 

Yes it is really original .ndpi, but I do not know why it show .tiff  I also downloaded it and it show .tif / but my friend also downloaded it and it show .ndpi extension maybe it is related google drive

Yeah I realized the difference, I have not encountered this before 

Please look at the [this ](https://streamable.com/uzty1p) video, and see I am flipping between them

And I also wonder why there is color differencess when I increase gamma value in Qupath0.5.0 see this pictures original one is more dark according to converted one. 


![jpeg_comp](https://github.com/libvips/libvips/assets/148692535/7aa98486-90b0-4b64-8f01-511fd6be88df)
![original_3](https://github.com/libvips/libvips/assets/148692535/a47e0740-d892-4304-af04-0cb5adf33b4b)

**Reply by jcupitt** [reply]
Sorry, I'm not sure what you are comparing. If you change the gamma you will see a colour difference.

**Comment by alexlsa** [comment]
Sorry maybe I could not tell my problem

There is `.ndpi` image, I convert it to `.tif` with `tiffsave`


After conversion finished, there is some color difference `between original and converted one`. I really see difference between them 

Yes If I change the gamma value of only one image, I will see a colour difference, but I am doing this operation between `original .ndpi` and `.tif(converted)` image. 

If there is no color difference between `.ndpi` and `.tif` image, I will not see any color difference between `.ndpi` and `.tif` when I change gamma value right ? (some compression artifact can be happen) 

I changed gamma value to see color difference more precisely. Actually when looking very carefully, I see difference between .ndpi and .tif image without changing the gamma value

Please see this images

`.tif` image compressed with `jpeg`. There are really brightness/color differences between `.ndpi and .tif image`. Maybe I am wrong somewhere or I could not tell my problem sory for that



![original_3_text](https://github.com/libvips/libvips/assets/148692535/d0d7ec16-4e5f-4737-8163-dbe04df0211e)
![jpeg_comp_text](https://github.com/libvips/libvips/assets/148692535/217f944c-9cea-4b0a-9c95-30204e0cebaa)

**Reply by jcupitt** [reply]
> After conversion finished, there is some color difference between original and converted one. I really see difference between them

Yes, but what exactly are you comparing? I need to understand the two paths the image is taking.

I think you might be comparing:

1. NDPI image -> tiffsave (uses openslide4 loader) -> QuPath (uses the tiff loader)
2. NDPI image -> QuPath (uses the bioformats loader, possibly a plain tiff loader, it's unclear, and depends on the qupath version)

Is that right? Or are you using the vendor viewer?

If that is right, the difference here is the bioformats NDPI loader vs. the openslide NDPI loader. I would expect them to be slightly different, and I would expect the openslide one to be more correct (or less incorrect haha). I think QuPath has an option to load via openslide, so that might be worth looking at. You'd need to ask on the qupath forums.

**Reply by alexlsa** [reply]
No, I am not using vendor viewer


I compare `.ndpi and .tif` image in qupath or another wsi viewer. Qupath also loaded .ndpi and .tif with `openslide` 

Actually I tried different viewer like [PMA](https://www.pathomation.com/pma-start/) and  [ASAP](https://computationalpathologygroup.github.io/ASAP/) still I see difference

I will also ask this in qupath forums

I think there is an issue inside `.ndpi image` I do not know

This issue **is not happen** for .svs, .mrxs .. now only for .ndpi images 


Thank you for your help


[video ](https://streamable.com/2czsaz)


![tiffsave](https://github.com/libvips/libvips/assets/148692535/d733124d-2fa3-4d2c-a638-3137c8dbe227)
![original](https://github.com/libvips/libvips/assets/148692535/ec792da8-0672-46f9-9ea4-c7370bb07d54)

**Reply by jcupitt** [reply]
I'd try vipsdisp as well -- it's 100% built against openslide4 and should be correct.

**Reply by alexlsa** [reply]
Hi @jcupitt 

Actually I found what is happening

There is no any issue about `openslide` and `vips`

When playing with .ndpi image **I realized that .ndpi image pyramid has different colors at some level**, please look at the I [shared](https://drive.google.com/file/d/1aInyeJ6e6CX7IYmOVDZrUp0vwlinfZP0/view) WSI in qupath at 2.47x zoom, then change zoom to 3.67x then you will see color is changed. 

When doing this operation please set gamma to 5 because it help to recognize color changing (zoom factor depends on the image) Also please look at white area when doing this operation because it is more recognizeable


I know, vips reads WSI from first level(by default) when doing tiffsave operation, that's why I think there is color issue between .ndpi and .tif

I downloaded openslide example `.ndpi` image and convert them to `.tif`  there is no color issue.

This happened for some .ndpi image

Please check that I will be sure but I think this is the issue

**Comment by alexlsa** [comment]
Hello @jcupitt,

Did you have a chance to look at it ? 

Thanks
